huaweicloud/services/modelarts/resource_huaweicloud_modelarts_authorization.go

```diff
@@ -0,0 +1,288 @@
+	"strings"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	"github.com/jmespath/go-jmespath"
+
+	"github.com/chnsz/golangsdk"
+	"github.com/chnsz/golangsdk/pagination"
+
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/common"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/config"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/utils"
+)
+
+func ResourceModelArtsAuthorization() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceModelArtsAuthorizationCreate,
+		UpdateContext: resourceModelArtsAuthorizationCreate,
+		ReadContext:   resourceModelArtsAuthorizationRead,
+		DeleteContext: resourceModelArtsAuthorizationDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"region": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+			},
+			"user_id": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: `User ID.`,
+			},
+			"type": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: `Authorization type. The valid value is **agency**.`,
+			},
+			"agency_name": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: `Agency name.`,
+				ValidateFunc: validation.StringMatch(regexp.MustCompile(`^(modelarts_agency)$|^(ma_agency_.+)$`),
+					"The agency name can be modelarts_agency or prefixed with ma_agency_"),
+			},
+			"user_name": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: `User Name.`,
+			},
+		},
+	}
+}
+
+func resourceModelArtsAuthorizationCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+	region := cfg.GetRegion(d)
+
+	// createAgency: create a ModelArts agency.
+	var (
+		createAgencyHttpUrl = "v2/{project_id}/agency"
+		createAgencyProduct = "modelarts"
+	)
+	createAgencyClient, err := cfg.NewServiceClient(createAgencyProduct, region)
+	if err != nil {
+		return diag.Errorf("error creating ModelArts Client: %s", err)
+	}
+
+	createAgencyPath := createAgencyClient.Endpoint + createAgencyHttpUrl
+	createAgencyPath = strings.ReplaceAll(createAgencyPath, "{project_id}", createAgencyClient.ProjectID)
+
+	createAgencyOpt := golangsdk.RequestOpts{
+		KeepResponseBody: true,
+		OkCodes: []int{
+			200,
+		},
+		MoreHeaders: map[string]string{"Content-Type": "application/json"},
+	}
+
+	createAgencyOpt.JSONBody = utils.RemoveNil(buildCreateAgencyBodyParams(d))
+	_, err = createAgencyClient.Request("POST", createAgencyPath, &createAgencyOpt)
+	if err != nil && parseAgencyErrorIgnoreConflict(err) != nil {
```

Comment:
The err from parseAgencyErrorIgnoreConflict(err) will be ignored, it is better to print it.